Search: Fix word exclusion in client side search

cglukas opened the pull request:
## Purpose

<!--
A description of the purpose of this pull request.
Ensure that all relevant information is included for reviewers,
including any environment-specific details.

* If you plan to add tests or documentation after opening this PR,
  please note it here.
* For user-visible changes, remember to add an entry to CHANGES.rst.
* Please add your name to AUTHORS.rst if you haven't already!
-->

The built-in search is capable of excluding search terms. Thats a great feature which would make the search a lot better!
Unfortunately, there are two blocking components:
- The `splitQuery` will discard hyphens which define the excluded terms
- The `performTermsSearch` will abort the search if any excluded term is matched

## References

Closes #13892

cglukas committed:
Fix that negated search words are filtered out in the splitQuery.

cglukas committed:
Fix that negated search words create a SyntaxError: Invalid or unexpected token.

cglukas committed:
Fix that excluded words would abort the entire search if matched in one page.

cglukas committed:
Fix format and add contribution information.

cglukas committed:
Fix long line.

cglukas commented on tests/js/searchtools.spec.js:
```diff
@@ -66,6 +66,32 @@ describe("Basic html theme search", function () {
     });
 
+    it("should find results when excluded terms are used", function () {
+      // This fixture already existed and has the right data to make this test work.
```
I don't know how to add new fixtures. JS is just a side hustle ^^. If you can give me a hint on how to add a custom fixture for that use case. I would be glad to integrate that too.

jayaddison replied:
The custom fixtures are built from content in the `tests/js/roots` directory -- each subdirectory there is a self-contained Sphinx project with a `conf.py` file.

It should be possible to run a Python utility script to generate/regenerate the fixture data from those projects, like this:

```sh
$ python utils/generate_js_fixtures.py
```

This will regenerate the fixture files, including any newly-added roots that you have added, and using the local version of Sphinx (it can be useful, for example, to activate your development branch in a `virtualenv` and then run that script if you are changing anything about the search index format -- in this PR we are not).

cglukas replied:
Hi @jayaddison,
thanks a lot for your hint. I'll check it out and update the PR accordingly.  Sounds very straight forward.

cglukas commented:
Regarding the CI: I don't see any changes on my PR which would trigger the current CI fail. Is this a common issue? TBH it does not look like it's affecting the master branch too 🤔. I'm a little aimless what to do here.

jayaddison commented:
> Regarding the CI: I don't see any changes on my PR which would trigger the current CI fail. Is this a common issue? TBH it does not look like it's affecting the master branch too 🤔. I'm a little aimless what to do here.

That's OK, yep - I believe that is due to bug #13886 (in progress, potentially to be fixed by #13883).

cglukas committed:
Add a dedicated fixture for testing excluded words in searches.

cglukas committed:
Is the pipeline flaky?

cglukas committed:
Add missing file.

cglukas committed:
Remove debug change.

AA-Turner committed:
Merge branch 'master' into work/fix-excluded-search-terms

# Conflicts:
#	CHANGES.rst

cglukas committed:
Reduce line length to fix doclinter.

cglukas committed:
Merge remote-tracking branch 'origin/master' into work/fix-excluded-search-terms

# Conflicts:
#	CHANGES.rst

cglukas committed:
Add regression test for TypeError in excluded words comparison.

cglukas committed:
Fix wording of test.

cglukas committed:
Update searchindex by running `python .\utils\generate_js_fixtures.py`

No change in functionality intended.
Only update the automatically generated files which might be forgotten.

cglukas committed:
Simplify regex code

Use lookbehind to split words only if a hyphon is contained inside a word.

Co-authored-by: James Addison <55152140+jayaddison@users.noreply.github.com>

cglukas committed:
Merge branch 'master' into work/fix-excluded-search-terms

cglukas committed:
Merge branch 'refs/heads/master' into work/fix-excluded-search-terms

cglukas committed:
Fxi searchindex.js of fixtures by rerunning generation with the newest updates from main.

cglukas committed:
Test commit to re-trigger the actions on github.

cglukas committed:
Test commit to re-trigger the actions on github.

cglukas committed:
Revert "Test commit to re-trigger the actions on github."

This reverts commit f1e5e8bc789e413164e418ec179204cb315d3ee6.
